### ISSUE TITLE
Neon sometimes renders the same neume component several times on top of each other.

### ISSUE THREAD

**Issue by cadagong** [DDMAL/Rodan#955]
This may be an MEI encoding issue, which is why I didn't raise it in the Neon repo. 
It's possible that the bug was introduced in the same manner as [this](https://github.com/DDMAL/Rodan/issues/802#issuecomment-1579283915) issue.

**Comment by cadagong** [comment]
Here's a screen recording of the issue:

https://github.com/DDMAL/Rodan/assets/40658078/5161a27a-7152-413f-88fa-e9cbc5e0b1c8

**Comment by fujinaga** [comment]
Please post the corresponding MEI file.

**Comment by cadagong** [comment]
Github doesn't allow me to add MEI files to comments here, but here is a link to it in my Google Drive:
https://drive.google.com/file/d/1uBzd1tJgc-HWBIuMANcFAm3MXTxMC4v4/view?usp=sharing

**Comment by cadagong** [comment]
Also please note that the folio had a lot more of these duplicated neume components all over the place.
If anyone wants to do some more digging around, here is a link to the full MEI file (Github doesn't allow me to upload it inside of a comment, so I uploaded it to my Google Drive):
https://drive.google.com/file/d/1uBzd1tJgc-HWBIuMANcFAm3MXTxMC4v4/view?usp=sharing

And here is the corresponding image of the folio:
![75](https://github.com/DDMAL/Rodan/assets/40658078/113967ca-513b-430d-ad2b-a281b8a6bee2)

**Comment by MrMondrian** [comment]
I ran a folio with the same problem in my E2E. The output of IC returns 136 glyphs and pitch finding finds 134. I think this may be interactive classifier. I agree that it isn't mei encoding

**Comment by sabrina0822** [comment]
I agree, I inspected the IC output for this folio and there seem to be some repeated elements with the exact same coordinates. There doesn't seem to be any necessary repeated behavior in terms of frequency or the types of elements that are repeated. Here are three examples: 

This clef appears 5 times in the IC output. 
```
<glyph uly="1586" ulx="897" nrows="129" ncols="36">
	<ids state="MANUAL">
		<id name="clef.c" confidence="1.000000"/>
	</ids>
```

This appears 5 times
```
<glyph uly="2772" ulx="1502" nrows="165" ncols="87">
	<ids state="MANUAL">
		<id name="neume.scandicus22b" confidence="1.000000"/>
	</ids>
```

This appears 4 times. 
```
<glyph uly="2004" ulx="2514" nrows="76" ncols="42">
	<ids state="MANUAL">
		<id name="custos" confidence="1.000000"/>
	</ids>
```

All of the repeats are being classified manually (at least from what I've seen so far) so I'm unsure what exactly is causing this problem.

**Comment by MrMondrian** [comment]
From looking at the output of IC both manual and automatically classified glyphs can become duplicates

**Comment by MrMondrian** [comment]
The source of this issue was in the function add_grouped_glyphs in [interactive_classifier.py](https://github.com/DDMAL/Rodan/blob/master/rodan-main/code/rodan/jobs/interactive_classifier/interactive_classifier.py). When glyphs are group together, and a server call is made, the client sends the grouped glyphs and then the server adds them. However if you don't close and reopen between servers calls, the same grouped glyphs keep getting sent. add_grouped_glyphs keeps re-adding the glyphs.

I was able to fix this locally by adding a check in add_grouped_glyphs that avoids adding a grouped glyph if a glyph with that id already exists. This solution involves an O(n^2) search, so maybe it isn't the best solution. It may be better for the client to not send the grouped glyphs more than once in the first place.

What bothers me is that this issue was too simple to not have always been there and never fixed. Futhermore, the python code for interactive classifier in the rodan repo has not been touched for a long time. A possible explanation is that the behaviour of gamera changed from version 3 to 4.

**Comment by fujinaga** [comment]
Great job finding the source of the error!

**Comment by sabrina0822** [comment]
**This only seems to happen when you save your work twice and then press re-group and reclassify. I think we need to empty out the dictionaries used in IC once your work has been saved to avoid duplicates.

**Comment by malajvan** [comment]
I've also been testing around with grouping and savings. Seems like everytime you save, instead of saving the CHANGES, it will just ADD any work that's been done onto the lists (i.e if you group a glyph, pressing save twice would literally group the glyph and add it to the list every single time you save). I agree with @sabrina0822 that we have to clear the "cache" everytime we save, A possible route is clearing everything on the backend and just redo the grouping operations every single time you press save.

**Comment by sabrina0822** [comment]
From testing a bit more it seems like this issue can be resolved if the window closes automatically every time you save the data. When I experiment saving and closing no repeats occur regardless of the other actions I do after. Is this a sufficient solution @timothydereuse or would it be better to manually clear out the data that get's passsed into the task every time you save? Not sure how easy the second solution would be but I could look into it.

**Comment by sabrina0822** [comment]
@JoyfulGen @martha-thomae from a user's perspective would it be annoying for Rodan to close interactive classifier when you use the save button and then to have to manually re-open it?

**Comment by malajvan** [comment]
So I guess our theory that it keeps sending the same glyphs and closing the tab clears the info stored by the browser. If we didn't want to close the tab everytime we save, maybe we could clear stuff everytime we save on the browser (does that make sense?) or some how get IC sends only the changes back to Rodan backend. If closing the tab everytime we save doesn't affect users' experience too much I think it is the quickest and cleanest fix.

**Comment by fujinaga** [comment]
Closing the IC every time one saves doesn't sounds like a great UI solution; but if that's the easiest, can you implement that for now and see if that, in fact, solves the duplicating issue?

**Comment by JoyfulGen** [comment]
> @JoyfulGen @martha-thomae from a user's perspective would it be annoying for Rodan to close interactive classifier when you use the save button and then to have to manually re-open it?

@sabrina0822 I do think this would be slightly annoying in the long run, especially if the user isn't warned it's going to happen. I'll also mention, just in case, that I'd never had this problem before. I was extensively using the IC until January 2022, used it again in May 2022, and then stopped actively using it from then on.

**Comment by sabrina0822** [comment]
Ok I'm going to work with someone who knows javascript better than I do today to try to find a better solution as opposed to just having the window close. 

Not sure why this is a new issue @JoyfulGen. When you were using it were you saving it multiple times before grouping and re-classifying?

**Comment by MrMondrian** [comment]
It seems that this is now the fourth really bad bug that was introduced during the push to production last summer/switch from py2->3. The other three being [this](https://github.com/DDMAL/Rodan/issues/860), [this](https://github.com/DDMAL/Rodan/issues/802), and [this](https://github.com/DDMAL/Rodan/issues/944). 

@sabrina0822 and @malajvan figured out how the IC issue was introduced, and I was able to figure out how the text alignment bug was introduced but have absolutely 0 clue where the hpf bug came from. 

I've looked everywhere for clues, even old docker images but have had no luck. It might be worthwhile to have some people help me dig around to see how the hpf and IC duplicate pitches bugs were introduced. I think it is important to figure this out because these two bugs seems to have always been in the code base, they just didn't present themselves until last summer. Something must've changed that we missed.

**Comment by JoyfulGen** [comment]
> Not sure why this is a new issue @JoyfulGen. When you were using it were you saving it multiple times before grouping and re-classifying?

That's actually a good question... It was a while ago, so I really can't be sure, but I don't think I was saving very much. I was probably just working until I was satisfied, and then re-grouping and going again. Sorry for the unhelpful reply!

**Comment by yinanazhou** [comment]
I'll close this issue since the changes are already merged.
